adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,496 @@
+is analogous to the work-group barrier.
+
+CUDA's warp concept has no SYCL equivalent. If a user were to write 
+warp aware code it would be non-generic SYCL code and specific to the 
+CUDA backend.
+
+CUDA allows for more detailed thread and memory management through 
+Cooperative Groups. Cooperative Groups allow for synchronizing at the 
+grid level and organizing subgroups in sizes smaller than a warp. 
+Cooperative Groups do not have an equivalent within SYCL 2020 and are 
+not yet supported.
+
+[[table.cuda.CUDA_features_to_SYCL]]
+.CUDA execution features with their corresponding SYCL features
+[width="100%",options="header",cols="50%,50%"]
+|====
+| [code]#SYCL#                                                       | [code]#CUDA#
+| [code]#Compute unit#                                               | [code]#Streaming multiprocessor#
+| [code]#nd_range#                                                   | [code]#grid#
+| [code]#work-group#                                                 | [code]#Thread block#
+| [code]#sub-group#                                                  | [code]#thread_block_tile#
+| [code]#work-item#                                                  | [code]#Thread#
+| [code]#SYCL nd_item synchronization#                               | [code]#cudaStreamSynchronize#
+| [code]#work-group barrier#                                         | [code]#__syncthread#
+|====
+
+[[sec::programming_interface]]
+== Programming Interface
+
+[[sub:cuda:application_interoperability]]
+=== Application Interoperability
+
+This section describes the API level interoperability between SYCL and CUDA.
+
+The CUDA backend supports API interoperability for `platform`, `device`,
+`context`, `queue`, `event` and `buffer`. Interoperability for `kernel`,
+`kernel_bundle`, `device_image`, `sampled_image` and `unsampled_image` is not
+supported.
+
+[[table.cuda.appinterop.nativeobjects]]
+.Types of native backend objects application interoperability
+[width="100%",options="header",cols="20%,20%,20%,40%"]
+|====
+| [code]#SyclType# | [code]#backend_input_t<backend::cuda, SyclType># | [code]#backend_return_t<backend::cuda, SyclType># | Description
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts , however it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue encapsulates a CUDA stream.
+| [code]#event#    | `CUevent`    | `CUevent`  | A SYCL event encapsulates a CUDA event.
+| [code]#buffer# | `struct { CUdeviceptr ptr; size_t size; }` | `CUdeviceptr` | A SYCL buffer encapsulates a CUDA device pointer.
```

Comment:
A question came up recently about the return value of `get_native()` for `buffer`.  What value is returned if `buffer` is a sub-buffer?  Is it a pointer to the original (non-sub) buffer, or is it an offset to the start of the sub-buffer?  It would be nice if there was a table describing the behavior of `get_native()` for each of these objects.  That table would be a good place to clarify things like this.

Reply:
Another question about the interop for `buffer` ... the OpenCL interop has a *vector* of `cl_mem` object because there might be multiple OpenCL devices on the system and the buffer might be represented on more than one device.  (At least I think that's the reasoning.)  Is it correct that there's just a single `CUdeviceptr` here, or should there be a vector of them?

Reply:
So we've been discussing application interop for `buffer` as well and for a few reasons we came to the conclusion that we should not support it and instead encourage `buffer` interop via `interop_handle::get_native_mem`:
* As you mentioned, as a `buffer` can represent multiple devices it would need to return a vector of `CUdeviceptr`, but because it allocates data lazily, this would also mean we would either need to return a partial list of memory objects depending on the current state of the runtime or enforce allocation within the `get_native`.
* If the return value has multiple memory objects, we would need some way to reflect which one has the latest copy of the data.
* A `buffer` can also represent memory on the host and potentially across multiple backends, and representing this would add further complexity to the return type.

In effect we came to the conclusion we would need to replicate much of the inner workings of a `buffer` in order to make it really useful to users, and the use cases we had could be supported via host task interop.